File: pypy/src/routes/notes.py
# testing chapter 5 database integration (partially untestable via http due to sqlite lock bugs)
from tina4_python.core.router import get, post, put, delete
from tina4_python.database.connection import Database


@get("/api/notes")
async def list_notes(request, response):
    db = Database()
    category = request.params.get("category")
    pinned = request.params.get("pinned")

    sql = "SELECT * FROM notes"
    params = []
    conditions = []

    if category:
        conditions.append("category = ?")
        params.append(category)

    if pinned == "true":
        conditions.append("pinned = 1")

    if conditions:
        sql += " WHERE " + " AND ".join(conditions)

    sql += " ORDER BY pinned DESC, created_at DESC"

    notes = db.fetch(sql, params)

    return response.json({"notes": notes, "count": len(notes)})


@get("/api/notes/categories")
async def list_categories(request, response):
    db = Database()
    categories = db.fetch(
        "SELECT category, COUNT(*) as count FROM notes GROUP BY category ORDER BY count DESC"
    )
    return response.json({"categories": categories})


@get("/api/notes/{id:int}")
async def get_note(request, response):
    db = Database()
    note = db.fetch_one(
        "SELECT * FROM notes WHERE id = ?",
        [request.params["id"]]
    )

    if note is None:
        return response.json({"error": "Note not found"}, 404)

    return response.json(note)


@post("/api/notes")
async def create_note(request, response):
    db = Database()
    body = request.body

    if not body.get("title"):
        return response.json({"error": "Title is required"}, 400)

    result = db.execute(
        "INSERT INTO notes (title, content, category, pinned) VALUES (?, ?, ?, ?)",
        [body["title"], body.get("content", ""), body.get("category", "general"), 1 if body.get("pinned") else 0]
    )

    note = db.fetch_one("SELECT * FROM notes WHERE id = ?", [result.last_id])
    db.cache_clear()

    return response.json({"message": "Note created", "note": note}, 201)


@put("/api/notes/{id:int}")
async def update_note(request, response):
    db = Database()
    note_id = request.params["id"]
    body = request.body

    existing = db.fetch_one("SELECT * FROM notes WHERE id = ?", [note_id])
    if existing is None:
        return response.json({"error": "Note not found"}, 404)

    db.execute(
        """UPDATE notes
           SET title = ?, content = ?, category = ?,
               pinned = ?, updated_at = CURRENT_TIMESTAMP
           WHERE id = ?""",
        [
            body.get("title", existing["title"]),
            body.get("content", existing["content"]),
            body.get("category", existing["category"]),
            1 if body.get("pinned", existing["pinned"]) else 0,
            note_id
        ]
    )

    updated = db.fetch_one("SELECT * FROM notes WHERE id = ?", [note_id])
    db.cache_clear()

    return response.json({"message": "Note updated", "note": updated})


@delete("/api/notes/{id:int}")
async def delete_note(request, response):
    db = Database()
    note_id = request.params["id"]

    existing = db.fetch_one("SELECT * FROM notes WHERE id = ?", [note_id])
    if existing is None:
        return response.json({"error": "Note not found"}, 404)

    db.execute("DELETE FROM notes WHERE id = ?", [note_id])
    db.cache_clear()

    return response.json(None, 204)

File: ruru/src/routes/notes.rb

 # notes crud api - ch05 exercise

# List all notes with optional filters
Tina4::Router.get("/api/notes") do |request, response|
  db = Tina4.database
  begin
    tag = request.params["tag"] || ""
    search = request.params["search"] || ""

    sql = "SELECT * FROM notes"
    params = []
    conditions = []

    unless tag.empty?
      conditions << "tag = ?"
      params << tag
    end

    unless search.empty?
      conditions << "(title LIKE ? OR content LIKE ?)"
      params << "%#{search}%"
      params << "%#{search}%"
    end

    sql += " WHERE #{conditions.join(' AND ')}" unless conditions.empty?
    sql += " ORDER BY updated_at DESC"

    notes = db.fetch(sql, params)

    response.json({
      notes: notes,
      count: notes.length
    })
  rescue => e
    response.json({ error: e.message }, 500)
  end
end

# Get a single note
Tina4::Router.get("/api/notes/{id:int}") do |request, response|
  db = Tina4.database
  begin
    id = request.params["id"]
    note = db.fetch_one("SELECT * FROM notes WHERE id = ?", [id])

    if note.nil?
      response.json({ error: "Note not found", id: id }, 404)
    else
      response.json(note)
    end
  rescue => e
    response.json({ error: e.message }, 500)
  end
end

# Create a note
Tina4::Router.post("/api/notes") do |request, response|
  db = Tina4.database
  begin
    body = request.body_parsed
    # validate
    errors = []
    errors << "Title is required" if body["title"].nil? || body["title"].empty?
    errors << "Content is required" if body["content"].nil? || body["content"].empty?

    unless errors.empty?
      next response.json({ errors: errors }, 400)
    end

    db.execute(
      "INSERT INTO notes (title, content, tag) VALUES (?, ?, ?)",
      [body["title"], body["content"], body["tag"] || "general"]
    )

    note = db.fetch_one("SELECT * FROM notes WHERE id = last_insert_rowid()")
    response.json(note, 201)
  rescue => e
    response.json({ error: e.message }, 500)
  end
end.no_auth

# Update a note
Tina4::Router.put("/api/notes/{id:int}") do |request, response|
  db = Tina4.database
  begin
    id = request.params["id"]
    body = request.body_parsed

    existing = db.fetch_one("SELECT * FROM notes WHERE id = ?", [id])
    if existing.nil?
      next response.json({ error: "Note not found", id: id }, 404)
    end

    db.execute(
      "UPDATE notes SET title = ?, content = ?, tag = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?",
      [
        body["title"] || existing["title"],
        body["content"] || existing["content"],
        body["tag"] || existing["tag"],
        id
      ]
    )

    note = db.fetch_one("SELECT * FROM notes WHERE id = ?", [id])
    response.json(note)
  rescue => e
    response.json({ error: e.message }, 500)
  end
end.no_auth

# Delete a note
Tina4::Router.delete("/api/notes/{id:int}") do |request, response|
  db = Tina4.database
  begin
    id = request.params["id"]
    existing = db.fetch_one("SELECT * FROM notes WHERE id = ?", [id])

    if existing.nil?
      next response.json({ error: "Note not found", id: id }, 404)
    end

    db.execute("DELETE FROM notes WHERE id = ?", [id])
    response.json(nil, 204)
  rescue => e
    response.json({ error: e.message }, 500)
  end
end.no_auth
